File: src/node/package/src/manual/native-bridge.ts
import { existsSync } from 'node:fs';
import * as path from 'node:path';
import koffi from 'koffi';
import { declareNativeFunctions, NativeFunctions } from '../generated/interop/native-functions';
import { QuestPdfNativeError } from './errors';

/**
 * The TypeScript half of the interop runtime. Generated bodies marshal their
 * arguments, call an export on [Native.lib], then [Native.check] for errors
 * recorded during the call (either a .NET exception reported through the error
 * callback, or a JavaScript exception thrown inside a user callback and
 * captured by [Native.guard]).
 *
 * Callback lifetime: koffi frees a callback's native trampoline when it is
 * unregistered, but the .NET side may hold the function pointer (a document's
 * content handler is invoked on every generate call), so every adapter is
 * registered through [Native.retain] and kept alive. Call
 * [Native.releaseRetainedCallbacks] between documents to reclaim them in
 * long-running processes.
 */

/** Error channel callback registered at initialization (see InteropRuntime.cs). */
const NativeErrorCallback = koffi.proto('void QuestPdfNativeErrorCallback(str message)');

interface RuntimeFunctions {
    releaseHandle: koffi.KoffiFunction;
    registerString: koffi.KoffiFunction;
    registerBuffer: koffi.KoffiFunction;
    freeString: koffi.KoffiFunction;
    freeBuffer: koffi.KoffiFunction;
    handleCount: koffi.KoffiFunction;
}

interface BridgeState {
    functions: NativeFunctions;
    runtime: RuntimeFunctions;
}

let state: BridgeState | null = null;
let pendingNativeError: string | null = null;
let pendingCallbackError: unknown = null;
const retained: koffi.IKoffiRegisteredCallback[] = [];

// The error channel must outlive releaseRetainedCallbacks(): the native side
// holds its function pointer for the whole process lifetime.
let errorCallbackRegistration: koffi.IKoffiRegisteredCallback | null = null;

/** The platform identifier used in @questpdf/native-* package names. */
function platformPackageId(): string {
    const arch = process.arch === 'arm64' ? 'arm64' : 'x64';

    if (process.platform === 'darwin')
        return `darwin-${arch}`;

    if (process.platform === 'win32')
        return `win32-${arch}`;

    return isMusl() ? `linux-musl-${arch}` : `linux-${arch}`;
}

/** glibc-based distributions report their glibc version in the process report; musl-based ones (Alpine) do not. */
function isMusl(): boolean {
    try {
        const report = process.report?.getReport() as { header?: { glibcVersionRuntime?: string } } | undefined;
        return report?.header?.glibcVersionRuntime === undefined;
    } catch {
        return false;
    }
}

function libraryDirectory(): string {
    const override = process.env.QUESTPDF_NATIVE_DIR;
    if (override !== undefined && override.length > 0)
        return path.resolve(override);

    // The installed platform package (@questpdf/native-*, an optional dependency of questpdf).
    try {
        const manifest = require.resolve(`@questpdf/native-${platformPackageId()}/package.json`);
        return path.join(path.dirname(manifest), 'native');
    } catch {
        // Fall through to the development layout.
    }

    // dist/manual/native-bridge.js → <package root>/native (development layout, npm run publish-native)
    return path.resolve(__dirname, '..', '..', 'native');
}

function libraryFileName(): string {
    switch (process.platform) {
        case 'darwin': return 'QuestPDF.Native.dylib';
        case 'win32': return 'QuestPDF.Native.dll';
        default: return 'QuestPDF.Native.so';
    }
}

function initialize(): BridgeState {
    const directory = libraryDirectory();
    const libraryPath = path.join(directory, libraryFileName());
    if (!existsSync(libraryPath)) {
        throw new Error(
            `Native library not found at ${libraryPath} — install the @questpdf/native-${platformPackageId()} ` +
            `package, point QUESTPDF_NATIVE_DIR at a published runtime directory, or run 'npm run publish-native' ` +
            `when working inside the repository.`);
    }

    const lib = koffi.load(libraryPath);

    // Handwritten lifecycle exports of QuestPDF.Native (see InteropRuntime.cs).
    const initializeExport = lib.func('QuestPdf_Initialize', 'int32_t', ['str', koffi.pointer(NativeErrorCallback)]);
    const runtime: RuntimeFunctions = {
        releaseHandle: lib.func('QuestPdf_ReleaseHandle', 'void', ['int64_t']),
        registerString: lib.func('QuestPdf_RegisterString', 'int64_t', ['str']),
        registerBuffer: lib.func('QuestPdf_RegisterBuffer', 'int64_t', [koffi.pointer('uint8_t'), 'int32_t']),
        freeString: lib.func('QuestPdf_FreeString', 'void', [koffi.pointer('uint8_t')]),
        freeBuffer: lib.func('QuestPdf_FreeBuffer', 'void', [koffi.pointer('uint8_t')]),
        handleCount: lib.func('QuestPdf_HandleCount', 'int64_t', []),
    };

    errorCallbackRegistration = koffi.register((message: string | null) => {
        pendingNativeError = message ?? 'Unknown native error';
    }, koffi.pointer(NativeErrorCallback));

    const rc = initializeExport(directory, errorCallbackRegistration) as number;
    if (rc !== 0)
        throw new Error(`QuestPdf_Initialize failed with code ${rc}`);

    return { functions: declareNativeFunctions(lib), runtime };
}

function ensure(): BridgeState {
    state ??= initialize();
    return state;
}

export const Native = {
    /** The declared native exports; loading and initialization happen on first access. */
    get lib(): NativeFunctions {
        return ensure().functions;
    },

    /** Rethrows any error recorded while the last native call was running. */
    check(): void {
        if (pendingCallbackError !== null) {
            const error = pendingCallbackError;
            pendingCallbackError = null;
            pendingNativeError = null;
            throw error;
        }

        if (pendingNativeError !== null) {
            const message = pendingNativeError;
            pendingNativeError = null;
            throw new QuestPdfNativeError(message);
        }
    },

    /** [check], then pass the value through — lets bridged bodies stay expressions. */
    checked<T>(value: T): T {
        Native.check();
        return value;
    },

    /**
     * Registers a JS function as a native callback of the given prototype and
     * keeps it strongly reachable so its trampoline stays valid: the .NET side
     * may hold the function pointer across calls.
     */
    retain(prototype: koffi.IKoffiCType, fn: (...args: never[]) => unknown): koffi.IKoffiRegisteredCallback {
        const registered = koffi.register(fn, koffi.pointer(prototype));
        retained.push(registered);
        return registered;
    },

    /**
     * Runs a user callback inside a native callback. A thrown exception must
     * not unwind through native frames, so it is recorded and rethrown by the
     * next [check]; the native side sees the fallback value.
     */
    guard<T>(fallback: T, block: () => T): T {
        try {
            return block();
        } catch (error) {
            pendingCallbackError ??= error;
            return fallback;
        }
    },

    /** Reads and frees a native UTF-8 string returned by an export. */
    takeString(pointer: unknown): string | null {
        if (pointer === null || pointer === undefined)
            return null;
        // 'char' with length -1 reads the NUL-terminated UTF-8 string at the
        // pointer ('str' would instead dereference it as a char* variable).
        const value = koffi.decode(pointer, 'char', -1) as string;
        ensure().runtime.freeString(pointer);
        return value;
    },

    /** Copies and frees a native buffer returned through out-parameters. */
    takeBuffer(pointer: unknown, length: number): Uint8Array {
        if (pointer === null || pointer === undefined || length === 0)
            return new Uint8Array(0);
        const bytes = koffi.decode(pointer, koffi.array('uint8_t', length, 'Typed')) as Uint8Array;
        ensure().runtime.freeBuffer(pointer);
        return bytes;
    },

    /** Registers a string as a transfer handle (callback results crossing to .NET). */
    registerString(value: string): number {
        return ensure().runtime.registerString(value) as number;
    },

    /** Registers a byte array as a transfer handle (callback results crossing to .NET). */
    registerBuffer(value: Uint8Array): number {
        return ensure().runtime.registerBuffer(value, value.length) as number;
    },

    /** Releases one handle of the native table (called by the finalization registry). */
    releaseHandle(handle: number): void {
        if (state !== null)
            state.runtime.releaseHandle(handle);
    },

    /** Number of live handles in the native table (diagnostics). */
    nativeHandleCount(): number {
        return ensure().runtime.handleCount() as number;
    },

    /**
     * Drops the registrations keeping callback trampolines alive. Only call
     * this when no document created so far will be generated again.
     */
    releaseRetainedCallbacks(): void {
        for (const callback of retained.splice(0))
            koffi.unregister(callback);
    },
};

File: src/jvm/package/src/manual/kotlin/com/questpdf/interop/NativeBridge.kt
package com.questpdf.interop

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.ptr.IntByReference
import com.sun.jna.ptr.PointerByReference
import java.io.File
import java.lang.ref.Cleaner
import java.nio.file.Files
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * The Kotlin half of the interop runtime. Generated bodies marshal their
 * arguments, call an export on [lib], then [check] for errors recorded during
 * the call (either a .NET exception reported through the error callback, or a
 * Kotlin exception thrown inside a user lambda and captured by [guard]).
 *
 * Callback lifetime: JNA frees a callback's native trampoline when the Kotlin
 * object is garbage collected, but the .NET side may hold the function pointer
 * (a document's content handler is invoked on every generate call), so every
 * adapter is [retain]ed. Call [releaseRetainedCallbacks] between documents to
 * reclaim them in long-running processes.
 */
object NativeBridge {

    private val retained = ConcurrentLinkedQueue<Any>()
    private val pendingNativeError = ThreadLocal<String?>()
    private val pendingCallbackError = ThreadLocal<Throwable?>()
    private val cleaner: Cleaner = Cleaner.create()

    internal val lib: QuestPdfNative by lazy(LazyThreadSafetyMode.SYNCHRONIZED) { load() }

    private fun load(): QuestPdfNative {
        System.setProperty("jna.encoding", "UTF-8")

        val directory = resolveNativeDirectory()
        val path = File(directory, nativeLibraryFileName())
        require(path.exists()) { "Native library not found at $path." }

        val loaded = Native.load(
            path.absolutePath,
            QuestPdfNative::class.java,
            mapOf(Library.OPTION_STRING_ENCODING to "UTF-8"),
        )

        val onError = NativeErrorCallback { message ->
            pendingNativeError.set(message ?: "Unknown native error")
        }
        retained.add(onError)

        val rc = loaded.QuestPdf_Initialize(directory.absolutePath, onError)
        check(rc == 0) { "QuestPdf_Initialize failed with code $rc" }

        registerApplicationFonts(loaded)

        return loaded
    }

    /**
     * Registers fonts deployed on the application classpath — the JVM analogue
     * of dropping font files into a .NET publish folder. Any jar may contribute
     * fonts by shipping a questpdf/fonts/index.txt resource (one path per line,
     * relative to questpdf/fonts/) next to the font files themselves; every
     * index on the classpath is honored, like META-INF/services registrations.
     */
    private fun registerApplicationFonts(loaded: QuestPdfNative) {
        val resourceRoot = "questpdf/fonts"
        val loader = Thread.currentThread().contextClassLoader ?: NativeBridge::class.java.classLoader

        val entries = buildList {
            for (index in loader.getResources("$resourceRoot/index.txt"))
                index.openStream().use { stream -> addAll(stream.bufferedReader().readLines()) }
        }
            .map(String::trim)
            .filter { it.isNotEmpty() }
            .distinct()

        if (entries.isEmpty())
            return

        val target = Files.createTempDirectory("questpdf-fonts-").toFile()
        Runtime.getRuntime().addShutdownHook(Thread { target.deleteRecursively() })

        for (entry in entries) {
            require(!entry.contains("..")) { "Invalid font index entry: $entry" }

            val destination = File(target, entry)
            destination.parentFile.mkdirs()

            val stream = loader.getResourceAsStream("$resourceRoot/$entry")
                ?: error("The font index lists $entry, but resource $resourceRoot/$entry is missing.")

            stream.use { input -> destination.outputStream().use { input.copyTo(it) } }
        }

        loaded.QuestPdf_RegisterFontDirectory(target.absolutePath)
        check()
    }

    /**
     * Locates the directory holding the QuestPDF.Native runtime files. An
     * explicit override (the questpdf.native.dir system property or the
     * QUESTPDF_NATIVE_DIR environment variable) wins; otherwise the files are
     * extracted from the natives classifier jar found on the classpath.
     */
    private fun resolveNativeDirectory(): File {
        val override = System.getProperty("questpdf.native.dir") ?: System.getenv("QUESTPDF_NATIVE_DIR")

        if (override != null) {
            val directory = File(override)
            require(directory.isDirectory) {
                "The QuestPDF native directory override points at $directory, which does not exist."
            }
            return directory
        }

        return extractNativesFromClasspath()
    }

    private fun nativeLibraryFileName(): String {
        val osName = System.getProperty("os.name").lowercase()
        return when {
            osName.contains("mac") -> "QuestPDF.Native.dylib"
            osName.contains("win") -> "QuestPDF.Native.dll"
            else -> "QuestPDF.Native.so"
        }
    }

    /** The platform identifier used by both the native build and the natives jar layout. */
    private fun currentRid(): String {
        val osName = System.getProperty("os.name").lowercase()
        val archName = System.getProperty("os.arch").lowercase()
        val arch = if (archName == "aarch64" || archName == "arm64") "arm64" else "x64"

        return when {
            osName.contains("mac") -> "osx-$arch"
            osName.contains("win") -> "win-$arch"
            isMusl() -> "linux-musl-$arch"
            else -> "linux-$arch"
        }
    }

    private fun isMusl(): Boolean =
        File("/etc/alpine-release").exists() ||
            File("/lib").listFiles().orEmpty().any { it.name.startsWith("ld-musl-") }

    /**
     * Extracts the runtime files listed in the natives jar's index into a
     * fresh temporary directory. Streams are read through the classloader, so
     * the mechanism works from plain classpath jars as well as repackaged
     * fat/nested jars (Spring Boot and friends).
     */
    private fun extractNativesFromClasspath(): File {
        val rid = currentRid()
        val resourceRoot = "questpdf/native/$rid"
        val loader = Thread.currentThread().contextClassLoader ?: NativeBridge::class.java.classLoader

        val entries = loader.getResourceAsStream("$resourceRoot/index.txt")?.use { stream ->
            stream.bufferedReader().readLines().map(String::trim).filter { it.isNotEmpty() }
        } ?: error(
            "QuestPDF native libraries for $rid were not found on the classpath. Add the natives " +
                "artifact for your platform (com.questpdf:questpdf with the '$rid' classifier), or point " +
                "the questpdf.native.dir system property / QUESTPDF_NATIVE_DIR environment variable at a " +
                "directory containing the published QuestPDF.Native runtime files."
        )

        val target = Files.createTempDirectory("questpdf-native-").toFile()
        Runtime.getRuntime().addShutdownHook(Thread { target.deleteRecursively() })

        for (entry in entries) {
            require(!entry.contains("..")) { "Invalid natives index entry: $entry" }

            val destination = File(target, entry)
            destination.parentFile.mkdirs()

            val stream = loader.getResourceAsStream("$resourceRoot/$entry")
                ?: error("The natives index lists $entry, but resource $resourceRoot/$entry is missing.")

            stream.use { input -> destination.outputStream().use { input.copyTo(it) } }
        }

        return target
    }

    /** Rethrows any error recorded while the last native call was running. */
    fun check() {
        val callbackError = pendingCallbackError.get()
        if (callbackError != null) {
            pendingCallbackError.remove()
            pendingNativeError.remove()
            throw callbackError
        }

        val nativeError = pendingNativeError.get()
        if (nativeError != null) {
            pendingNativeError.remove()
            throw QuestPdfNativeException(nativeError)
        }
    }

    /** [check], then pass the value through — lets bridged bodies stay expressions. */
    fun <T> checked(value: T): T {
        check()
        return value
    }

    /** Keeps a JNA callback strongly reachable so its native trampoline stays valid. */
    internal fun <T : Any> retain(callback: T): T {
        retained.add(callback)
        return callback
    }

    /**
     * Runs a user lambda inside a native callback. A thrown exception must not
     * unwind through native frames, so it is recorded and rethrown by the next
     * [check] on the Kotlin side; the native side sees the fallback value.
     */
    fun <T> guard(fallback: T, block: () -> T): T =
        try {
            block()
        } catch (t: Throwable) {
            if (pendingCallbackError.get() == null)
                pendingCallbackError.set(t)
            fallback
        }

    /** Reads and frees a native UTF-8 string returned by an export. */
    fun takeString(pointer: Pointer?): String? {
        if (pointer == null)
            return null
        val value = pointer.getString(0, "UTF-8")
        lib.QuestPdf_FreeString(pointer)
        return value
    }

    /** Copies and frees a native buffer returned through out-parameters. */
    fun takeBuffer(data: PointerByReference, length: IntByReference): ByteArray {
        val pointer = data.value ?: return ByteArray(0)
        val bytes = pointer.getByteArray(0, length.value)
        lib.QuestPdf_FreeBuffer(pointer)
        return bytes
    }

    /** Registers a string as a transfer handle (callback results crossing to .NET). */
    fun registerString(value: String): Long = lib.QuestPdf_RegisterString(value)

    /** Registers a byte array as a transfer handle (callback results crossing to .NET). */
    fun registerBuffer(value: ByteArray): Long = lib.QuestPdf_RegisterBuffer(value, value.size)

    /** Number of live handles in the native table (diagnostics). */
    fun nativeHandleCount(): Long = lib.QuestPdf_HandleCount()

    /**
     * Drops the strong references keeping callback trampolines alive. Only call
     * this when no document created so far will be generated again.
     */
    fun releaseRetainedCallbacks() {
        retained.clear()
    }

    internal fun registerForCleanup(target: Any, handle: Long) {
        cleaner.register(target) { lib.QuestPdf_ReleaseHandle(handle) }
    }
}
